common/os.go

```diff
@@ -126,30 +126,42 @@ func MustWriteFile(filePath string, contents []byte, mode os.FileMode) {
 
 // WriteFileAtomic writes newBytes to temp and atomically moves to filePath
 // when everything else succeeds.
-func WriteFileAtomic(filePath string, newBytes []byte, mode os.FileMode) error {
+func WriteFileAtomic(filePath string, newBytes []byte, mode os.FileMode) (err error) {
 	dir := filepath.Dir(filePath)
 	f, err := ioutil.TempFile(dir, "")
 	if err != nil {
 		return err
 	}
-	_, err = f.Write(newBytes)
-	if err == nil {
-		err = f.Sync()
-	}
-	if closeErr := f.Close(); err == nil {
-		err = closeErr
+
+	defer func() {
+		// Any error should result in full cleanup.
+		// See Issue https://github.com/tendermint/tmlibs/issues/159
+		if err != nil {
+			// If the file happens to be open, let's close it
+			// lest be hung up on it such as on Windows
+			// where we need to first close before removing
+			// See Issue https://github.com/golang/go/issues/16296.
+			if f != nil {
```

Comment:
nice find! why just don't add this to 

```
// any err should result in full cleanup
if err != nil {
  <HERE>
  os.Remove(f.Name())	
}
```

Reply:
The goal here is to close if the file is non-nil before os.Remove. I should have instead moved `os.Remove` in the `if f != nil` check